### PR TITLE
MODINREACH-351: Ensure that ongoing contributions can retry when central server is unavailable

### PR DESCRIPTION
**Pull request by Arghya22**
MODINREACH-351  (https://issues.folio.org/browse/MODINREACH-351)

## Purpose
Ensure that ongoing contributions can retry when central server is unavailable.
 
## Approach
1. Blocking kafka consumer retry implemented for inventory listener 
2. New repository method created for not creating multiple contribution on each retry 
3. Code adjusted 
4. Test case implemented

### CODE REVIEW

**Review by pavankumar181 — APPROVED** [review]

**Review by Vignesh-kalyanasundaram — APPROVED** [review]

**Comment by steveellis** [comment]
@Arghya22 it would be greatly appreciated if you could create a PR with _only_ the changes from 151. This has all the changes from 347 too, making for a giant PR which mostly contains stuff I've already reviewed.

This should be easy since master now has 347. Just go back to before you merged 347 here and merge in master to clean things up.

Also, I'm noticing that the code doesn't have my formatting changes and log changes, so it looks like even this branch doesn't have the latest from 347.

File: src/main/java/org/folio/innreach/domain/service/impl/BatchDomainEventProcessor.java [hunk 1]
```diff
@@ -36,7 +39,12 @@ public <T> void process(List<DomainEvent<T>> batch, Consumer<DomainEvent<T>> rec
       try {
         executionService.runTenantScoped(tenantId,
           () -> processTenantEvents(events, recordProcessor));
-      } catch (ListenerExecutionFailedException | FeignException ex) {
+      }
+      catch (ServiceSuspendedException | FeignException | InnReachConnectionException | SocketTimeOutExceptionWrapper e) {
+        log.info("exception thrown from process");
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/BatchDomainEventProcessor.java]: Would be a good idea to log the actual exception thrown here (`log.info("Exception thrown from process", e)`) to make debugging in prod easier.

File: src/main/java/org/folio/innreach/domain/service/impl/BatchDomainEventProcessor.java [hunk 2]
```diff
@@ -47,11 +55,13 @@ private <T> void processTenantEvents(List<DomainEvent<T>> events, Consumer<Domai
...
       log.info("Processing event {}", event);
       try {
-        retryTemplate.execute(ctx -> {
           recordProcessor.accept(event);
-          return null;
-        });
-      } catch (Exception e) {
+      }
+      catch (ServiceSuspendedException | FeignException | InnReachConnectionException e) {
+        log.info("exception thrown from process");
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/BatchDomainEventProcessor.java]: Same as above. Let's log the exception so we can know which one of the three happened.

File: src/main/java/org/folio/innreach/domain/service/impl/ContributionActionServiceImpl.java [hunk 3]
```diff
@@ -119,13 +126,17 @@ public void handleItemDelete(Item deletedItem) {
     if (!isMARCRecord(instance)) {
       return;
     }
-
-    for (var csId : getCentralServerIds()) {
-      contributionJobRunner.runItemDeContribution(csId, instance, deletedItem);
+    try{
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/ContributionActionServiceImpl.java]: Need a space after `try`

File: src/main/java/org/folio/innreach/domain/service/impl/ContributionActionServiceImpl.java [hunk 4]
```diff
@@ -119,13 +126,17 @@ public void handleItemDelete(Item deletedItem) {
...
+      for (var csId : getCentralServerIds()) {
+        contributionJobRunner.runItemDeContribution(csId, instance, deletedItem);
+      }
+    } catch (ServiceSuspendedException | FeignException | InnReachConnectionException e) {
+      log.info("exception thrown from handleItemDelete");
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/ContributionActionServiceImpl.java]: Same as above. Log the exception so we know which one happened

File: src/main/java/org/folio/innreach/domain/service/impl/ContributionActionServiceImpl.java [hunk 5]
```diff
@@ -199,8 +210,14 @@ public void handleHoldingDelete(Holding holding) {
...
     var items = holding.getHoldingsItems();
-    for (var csId : getCentralServerIds()) {
-      items.forEach(item -> contributionJobRunner.runItemDeContribution(csId, instance, item));
+    try {
+      for (var csId : getCentralServerIds()) {
+        items.forEach(item -> contributionJobRunner.runItemDeContribution(csId, instance, item));
+      }
+    }
+    catch (ServiceSuspendedException | FeignException | InnReachConnectionException e) {
+      log.info("exception thrown from handleHoldingDelete");
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/ContributionActionServiceImpl.java]: Same as previous. Need to log the exception here.

File: src/main/java/org/folio/innreach/domain/service/impl/ContributionActionServiceImpl.java [hunk 6]
```diff
@@ -213,7 +230,12 @@ private void handlePerCentralServer(UUID recordId, Consumer<UUID> centralServerH
         } else {
           log.warn("Central server {} contribution configuration is not valid", csId);
         }
-      } catch (Exception e) {
+      }
+      catch (ServiceSuspendedException | FeignException | InnReachConnectionException | SocketTimeOutExceptionWrapper e) {
+        log.info("exception thrown from handlePerCentralServer");
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/ContributionActionServiceImpl.java]: Same as previous. Need to log the exception here.

File: src/test/java/org/folio/innreach/batch/contribution/service/ContributionJobRunnerTest.java [hunk 7]
```diff
@@ -87,108 +100,148 @@ class ContributionJobRunnerTest {
...
   @InjectMocks
   private ContributionJobRunner jobRunner;
 
+  @Mock
+  private InitialContributionJobConsumerContainer initialContributionJobConsumerContainer;
+
+  @Spy
+  private ConcurrentHashMap<String, Integer> recordsProcessed = new ConcurrentHashMap<>();
+
+  //added
```
Comment (steveellis) [inline src/test/java/org/folio/innreach/batch/contribution/service/ContributionJobRunnerTest.java]: Probably can remove this comment

**Comment by steveellis** [comment]
> @Arghya22 it would be greatly appreciated if you could create a PR with _only_ the changes from 151. This has all the changes from 347 too, making for a giant PR which mostly contains stuff I've already reviewed.
> 
> This should be easy since master now has 347. Just go back to before you merged 347 here and merge in master to clean things up.
> 
> Also, I'm noticing that the code doesn't have my formatting changes and log changes, so it looks like even this branch doesn't have the latest from 347.

I rebased the changes you have labeled in your commit messages out to another branch which made it easier to compare the changes. I think it is probably less work to keep your branch for merging. I think this looks good besides the minor things I highlight above.

**Comment by steveellis** [comment]
@Arghya22 Might as well keep this branch open until we do another round of testing. If that testing doesn't have any issues with ongoing I think we can consider it done.

**Review by steveellis — APPROVED** [review]
@Arghya22 - Thanks for all your work on this. I think based on the successful test today you can safely merge this in.

**Review by gurleenkaurbp — APPROVED** [review]

**Pull request merged**